torch/csrc/distributed/rpc/process_group_agent.h

```diff
@@ -70,13 +71,45 @@ class ProcessGroupAgent : public RpcAgent {
     std::mutex mutex_;
   };
 
+  // The FutureInfo struct stores a shared_ptr to the future, as well as
+  // additional information to manage timeouts and destination information,
+  // which is needed for termination detection.
+  struct FutureInfo {
+    std::shared_ptr<FutureMessage> future_;
+    std::chrono::milliseconds startTime_;
+    int dstRank_;
+    std::chrono::milliseconds timeout_;
+    FutureInfo(
+        const std::shared_ptr<FutureMessage>& future,
+        const std::chrono::milliseconds& startTime,
+        int dstRank,
+        const std::chrono::milliseconds timeout)
+        : future_(future),
+          startTime_(startTime),
+          dstRank_(dstRank),
+          timeout_(timeout) {}
+    FutureInfo() {}
```

Comment:
Beware: this allows for instances with garbage data. I'd either add member initializers here, or remove the default constructor entirely (if you're worried about inserting them into maps, use `std::move` or `emplace`).

Reply:
Fixed in https://github.com/pytorch/pytorch/pull/30197